### PR TITLE
Refactor Pusher#load_spec to use Gem::Package

### PR DESCRIPTION
**Pull request by arthurnn**
Gem::Package provide a way to load the spec from the .gem file, so the
pusher should use that. On rubygems 2.3+ that method start receiving any
IO and not just a file. As we are on that version, we can use that
method.

This changes one small behaviour though. Now the pusher will only be
able to load_spec for gems that include a data.tar.gz . Gem pushes
without a data will not be accepted anymore.

see https://github.com/rubygems/rubygems/blob/f24c244308266970fdfdee85f5cba8532c936b79/lib/rubygems/package.rb#L663-L666

as suggested by @drbrain https://github.com/rubygems/rubygems.org/pull/854#discussion_r23952451

review @qrush @dwradcliffe

### CODE REVIEW

File: test/unit/pusher_test.rb [hunk 1]
```diff
@@ -102,6 +100,13 @@ class PusherTest < ActiveSupport::TestCase
       assert_not_nil @cutter.spec.dependencies.first.requirement
     end
 
+    should "not be able to pull spec when no data available" do
+      @gem = gem_file("aliases-nodata-0.0.1.gem")
+      @cutter = Pusher.new(@user, @gem)
+      @cutter.pull_spec
+      assert_includes @cutter.message, %{package content (data.tar.gz) is missing}
+    end
```
Comment (arthurnn) [inline test/unit/pusher_test.rb]: the description of the PR, states this behaviour change.

**Comment by arthurnn** [comment]
any concerns about this? 
wondering to merge it this soon (maybe not today :smile: , tho)

**Comment by drbrain** [comment]
If you don't have a data.tar.gz RubyGems will also reject it so I think this is OK

**Comment by qrush** [comment]
@drbrain what versions of RG don't `gem build` with a `data.tar.gz`? Maybe we could use statsd to track how many we'd end up rejecting.

**Comment by arthurnn** [comment]
:mask:

**Comment by drbrain** [comment]
All versions with the tar format have a data.tar.gz. If you package fully by hand you could make a gem without a data.tar.gz, but we don't need to encourage that.

0.7 and older have a YAML-esque format, but nobody should be uploading them anymore.

**Pull request merged**